src/main/clojure/clara/rules/compiler.clj

```diff
@@ -235,15 +235,21 @@
   "Compiles a sequence of constraints into a structure that can be evaluated.
 
    Callers may also pass a collection of equality-only-variables, which instructs
-   this function to only do an equality check on them rather than create a unification binding."
+   this function to only do an equality check on them rather than create a unification binding.
+
+   The returned expression evaluates to a two item vector, [bindings error], which contains
+   the evaluated bindings in the first argument and a map with information on any exception
+   that might have been thrown by evaluating a constraint as the second vector entry."
```

Comment:
Regarding the cost of using the error/bindings tuples, that is a good point @rbrush and I agree.  Obviously the cost of instantiating and destructuring the vector is small but it could add up.  We've seen some very tiny optimizations in our rule constraints that end up being part of the AlphaNode filters in particular be noticeable before.

 I think the reason this was implemented this way is that it derives a lot of information about the context in which the exception was thrown - i.e. the entire condition, the productions(rules/queries) that the constraint is a part of, etc. by examining the rule network at the time the exception is thrown.  In principle this information could be determined at compilation time rather than re-derived but it looked to me like this would involve significant development work in the compiler and since this analysis is only done when an exception is thrown, and thus should not be a performance problem, I suggested that we stick with deriving the information at runtime.

However, on further examination this does not preclude the approach you suggest.  Take part of the code from the current PR:

````
(defn- alpha-node-matches
  [facts env activation node]
  (for [fact facts
        :let [[bindings error] (activation fact env)]
        :when (or bindings error)] ; FIXME: add env.
    (if error
      (throw-constraint-exception error node fact env)
      [fact bindings])))
````

The throw-constraint-exception function is called with error, node, fact, and env.  The contents of error are derived from fact and env by the activation call.  Therefore, if we called the activation with the node in addition to its existing arguments it could derive all this information itself and then throw the exception itself.  I don't think this would be too difficult to do; I would anticipate that we could just keep the existing infrastructure Carlos has built to gather the information, format the messages, etc. and just call it elsewhere i.e. in the compiled filter or join function.

Regarding throwing a specialized exception, are you suggesting that we implement our own exception class?  If so I suggest that it implement IExceptionInfo, which would not be too difficult to do.  I've found the ability to inspect data relevant to an exception programmatically at the REPL to be quite valuable.  However, I'm not sure this is something that adds much value.  It isn't obvious to me why a consumer would only want to catch such a specialized exception type as opposed to a more general exception type.  For example, obviously I'd hope this doesn't happen, but if Clara itself threw an exception due to a bug I'd think a consumer would want to handle it in much the same way as an exception from a function call in their rule constraints.  I'm not opposed to making our own type if we land on use-cases for it though, it just seems somewhat premature to me if we don't know of a need for it yet, especially since this would be platform-specific work i.e. different between Clojure and ClojureScript.

I agree that especially since Carlos has other commitments we should be willing to get this over the finish line.  This sort of error handling is likely to be very beneficial.  Regarding the procedural aspects of that I'm not quite sure what form that would take; I do want to be sure that the Git history ends up appropriately crediting Carlos for this significant piece of work.